README.md
[![awscdk-jsii-template](https://img.shields.io/badge/built%20with-awscdk--jsii--template-blue)](https://github.com/pahud/awscdk-jsii-template)
[![NPM version](https://badge.fury.io/js/cdk-spot-one.svg)](https://badge.fury.io/js/cdk-spot-one)
[![PyPI version](https://badge.fury.io/py/cdk-spot-one.svg)](https://badge.fury.io/py/cdk-spot-one)
![Release](https://github.com/pahud/cdk-spot-one/workflows/Release/badge.svg)

# cdk-spot-one

One spot instance with EIP and defined duration. No interruption.

# Why

Sometimes we need an Amazon EC2 instance with static fixed IP for testing or development purpose for a duration of
time(probably hours). We need to make sure during this time, no interruption will occur and we don't want to pay
for on-demand rate. `cdk-spot-one` helps you reserve one single spot instance with pre-allocated or new
Elastic IP addresses(EIP) with defined `blockDuration`, during which time the spot instance will be secured with no spot interruption.

Behind the scene, `cdk-spot-one` provisions a spot fleet with capacity of single instance for you and it associates the EIP with this instance. The spot fleet is reserved as spot block with `blockDuration` from one hour up to six hours to ensure the high availability for your spot instance.

Multiple spot instances are possible by simply specifying the `targetCapacity` construct property, but we only associate the EIP with the first spot instance at this moment.

Enjoy your highly durable one spot instance with AWS CDK!

# Sample

```ts
import { SpotFleet } from 'cdk-spot-one';

// create the first fleet for one hour and associate with our existing EIP
const fleet = new SpotFleet(stack, 'SpotFleet')

// configure the expiration after 1 hour
fleet.expireAfter(Duration.hours(1))


// create the 2nd fleet with single Gravition 2 instance for 6 hours and associate with new EIP
const fleet2 = new SpotFleet(stack, 'SpotFleet2', {
  blockDuration: BlockDuration.SIX_HOURS,
  eipAllocationId: 'eipalloc-0d1bc6d85895a5410',
  defaultInstanceType: new InstanceType('c6g.large'),
  vpc: fleet.vpc,
})
// configure the expiration after 6 hours
fleet2.expireAfter(Duration.hours(6))

// print the instanceId from each spot fleet
new CfnOutput(stack, 'SpotFleetInstanceId', { value: fleet.instanceId })
new CfnOutput(stack, 'SpotFleet2InstanceId', { value: fleet2.instanceId })

```

# Create spot instances without duration block

```ts
const fleet = new SpotFleet(stack, 'SpotFleet', {
  blockDuration: BlockDuration.NONE,
})
```

NOTE: This kind of spot instance will be interrupted by AWS. However the fleet is using type [maintain](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet.html#spot-fleet-allocation-strategy), the fleet can be refulfilled.

# ARM64 and Graviton 2 support

`cdk-spot-one` selects the latest Amazon Linux 2 AMI for your `ARM64` instances. Simply select the instance types with the `defaultInstanceType` property and the `SpotFleet` will auto configure correct AMI for the instance.


```ts
defaultInstanceType: new InstanceType('c6g.large')
```

# SSH connect

By default the `cdk-spot-one` does not bind any SSH public key for you on the instance. You are encouraged to use `ec2-instance-connect` to send your public key from local followed by one-time SSH connect.

For example:

```sh
pubkey="$HOME/.ssh/aws_2020_id_rsa.pub"
echo "sending public key to ${instanceId}"
aws ec2-instance-connect send-ssh-public-key --instance-id ${instanceId} --instance-os-user ec2-user \
--ssh-public-key file://${pubkey} --availability-zone ${az} 
```
## npx ec2-connect INSTANCE_ID

To connect to the instance, run `npx ec2-connect` as below:

```sh
$ npx ec2-connect i-01f827ab9de7b93a9
```

or 

```sh
$ npx ec2-connect i-01f827ab9de7b93a9 ~/.ssh/other_public_key_path
```
If you are using different SSH public key(default is ~/.ssh/id_rsa.pub)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.6.116 (2020-12-21)

### 0.6.115 (2020-12-20)

### 0.6.114 (2020-12-19)

### 0.6.113 (2020-12-18)

### 0.6.112 (2020-12-17)

### 0.6.111 (2020-12-16)

### 0.6.110 (2020-12-15)

### 0.6.109 (2020-12-14)

### 0.6.108 (2020-12-13)

### 0.6.107 (2020-12-12)

### 0.6.106 (2020-12-11)

### 0.6.105 (2020-12-10)

### 0.6.104 (2020-12-09)

### 0.6.103 (2020-12-08)

### 0.6.102 (2020-12-08)

### 0.6.101 (2020-12-08)

### 0.6.100 (2020-12-08)

### 0.6.99 (2020-11-24)

### 0.6.98 (2020-11-23)

### 0.6.97 (2020-11-21)

### 0.6.96 (2020-11-20)

### 0.6.95 (2020-11-19)

### 0.6.94 (2020-11-18)

### 0.6.93 (2020-11-17)

### 0.6.92 (2020-11-16)

### 0.6.91 (2020-11-15)

### 0.6.90 (2020-11-14)

### 0.6.89 (2020-11-13)

### 0.6.88 (2020-11-12)

### 0.6.87 (2020-11-11)

### 0.6.86 (2020-11-10)

### 0.6.85 (2020-11-09)

### 0.6.84 (2020-11-09)

### 0.6.83 (2020-11-09)

### 0.6.82 (2020-11-08)

### 0.6.81 (2020-11-07)

### 0.6.80 (2020-11-06)

### 0.6.79 (2020-11-05)

### 0.6.78 (2020-11-04)

### 0.6.77 (2020-11-03)

### 0.6.76 (2020-11-02)

### 0.6.75 (2020-11-01)

### 0.6.74 (2020-10-31)

### 0.6.73 (2020-10-30)

### 0.6.72 (2020-10-29)

### 0.6.71 (2020-10-28)

### 0.6.70 (2020-10-27)

### 0.6.69 (2020-10-26)

### 0.6.68 (2020-10-25)

### 0.6.67 (2020-10-24)

### 0.6.66 (2020-10-23)

### 0.6.65 (2020-10-22)

### 0.6.64 (2020-10-21)

### 0.6.63 (2020-10-19)

### 0.6.62 (2020-10-19)

### 0.6.61 (2020-10-13)

### 0.6.60 (2020-10-08)

### 0.6.59 (2020-10-08)

### 0.6.58 (2020-10-08)

### 0.6.57 (2020-10-08)

### 0.6.56 (2020-10-07)

### 0.6.55 (2020-10-07)

### 0.6.54 (2020-10-07)

### 0.6.53 (2020-10-07)

### 0.6.52 (2020-10-07)

### 0.6.51 (2020-10-07)

### 0.6.50 (2020-10-06)

### 0.6.49 (2020-10-06)

### 0.6.48 (2020-10-05)

### 0.6.47 (2020-10-05)

### 0.6.46 (2020-09-30)

### 0.6.45 (2020-09-30)

### 0.6.44 (2020-09-29)

### 0.6.43 (2020-09-29)

### 0.6.42 (2020-09-29)

### 0.6.41 (2020-09-28)

### 0.6.40 (2020-09-28)

### 0.6.39 (2020-09-27)

### 0.6.38 (2020-09-27)

### 0.6.37 (2020-09-27)

### 0.6.36 (2020-09-27)

### [0.6.35](https://github.com/pahud/cdk-spot-one/compare/v0.6.34...v0.6.35) (2020-09-25)


### Bug Fixes

* **eip-handler:** add verbose error message ([2995b62](https://github.com/pahud/cdk-spot-one/commit/2995b620a6a49334b1b2e285a99df6f2e99fb195))

### [0.6.34](https://github.com/pahud/cdk-spot-one/compare/v0.6.33...v0.6.34) (2020-09-09)

### 0.6.33 (2020-09-08)

### 0.6.32 (2020-09-08)


### Features

* **spot:** allow creating spot fleets with no block duration ([#39](https://github.com/pahud/cdk-spot-one/issues/39)) ([16d8713](https://github.com/pahud/cdk-spot-one/commit/16d871360eb0f56d783ee7b53970ce7fbd87ea0a)), closes [#38](https://github.com/pahud/cdk-spot-one/issues/38)

### 0.6.31 (2020-09-08)

### 0.6.30 (2020-09-08)


### Features

* support additional user data. ([#34](https://github.com/pahud/cdk-spot-one/issues/34)) ([a10a399](https://github.com/pahud/cdk-spot-one/commit/a10a3995cf746e718168d5b6a3e10f5fe1c27383)), closes [#35](https://github.com/pahud/cdk-spot-one/issues/35)

### [0.6.29](https://github.com/pahud/cdk-spot-one/compare/v0.6.28...v0.6.29) (2020-09-06)

### 0.6.28 (2020-09-04)

### [0.6.27](https://github.com/pahud/cdk-spot-one/compare/v0.6.26...v0.6.27) (2020-09-03)

### 0.6.26 (2020-09-03)

### 0.6.25 (2020-09-02)

### 0.6.24 (2020-09-01)

### [0.6.23](https://github.com/pahud/cdk-spot-one/compare/v0.6.21...v0.6.23) (2020-09-01)

### [0.6.22](https://github.com/pahud/cdk-spot-one/compare/v0.6.20...v0.6.22) (2020-08-31)

### 0.6.22 (2020-08-31)

### 0.6.21 (2020-08-31)

### [0.6.20](https://github.com/pahud/cdk-spot-one/compare/v0.6.19...v0.6.20) (2020-08-28)

### 0.6.19 (2020-08-28)

### 0.6.18 (2020-08-28)

### [0.6.17](https://github.com/pahud/cdk-spot-one/compare/v0.6.16...v0.6.17) (2020-08-27)

### [0.6.16](https://github.com/pahud/cdk-spot-one/compare/v0.6.15...v0.6.16) (2020-08-27)

### [0.6.15](https://github.com/pahud/cdk-spot-one/compare/v0.6.14...v0.6.15) (2020-08-27)

### [0.6.14](https://github.com/pahud/cdk-spot-one/compare/v0.6.13...v0.6.14) (2020-08-27)

### 0.6.13 (2020-08-27)

### 0.6.12 (2020-08-24)

### [0.6.11](https://github.com/pahud/cdk-spot-one/compare/v0.6.8...v0.6.11) (2020-08-24)

### [0.6.10](https://github.com/pahud/cdk-spot-one/compare/v0.6.7...v0.6.10) (2020-08-21)

### 0.6.10 (2020-08-21)

### 0.6.9 (2020-08-19)

### 0.6.8 (2020-08-15)

### [0.6.7](https://github.com/pahud/cdk-spot-one/compare/v0.6.5...v0.6.7) (2020-08-15)
### 0.6.6 (2020-08-15)

### [0.6.5](https://github.com/pahud/cdk-spot-one/compare/v0.6.4...v0.6.5) (2020-08-13)

### 0.6.4 (2020-08-08)

### 0.6.3 (2020-08-08)

### [0.6.2](https://github.com/pahud/cdk-spot-one/compare/v0.6.1...v0.6.2) (2020-08-08)


### Bug Fixes

* **core:** python module name typo ([2a3ddf9](https://github.com/pahud/cdk-spot-one/commit/2a3ddf9c614f3f25838a70ad1c092f25c50583a3)) closes [#3](https://github.com/pahud/cdk-spot-one/issues/3)

### [0.6.1](https://github.com/pahud/cdk-spot-one/compare/v0.6.0...v0.6.1) (2020-08-08)

## [0.6.0](https://github.com/pahud/cdk-spot-one/compare/v0.5.0...v0.6.0) (2020-08-08)

## [0.5.0](https://github.com/pahud/cdk-spot-one/compare/v0.4.3...v0.5.0) (2020-07-20)

### [0.4.3](https://github.com/pahud/cdk-spot-one/compare/v0.4.2...v0.4.3) (2020-07-20)


### Bug Fixes

* **handler:** failed to destroy stack ([77efe07](https://github.com/pahud/cdk-spot-one/commit/77efe0737229f2817f98514aeaff509c0c8a3bb6))

### [0.4.2](https://github.com/pahud/cdk-spot-one/compare/v0.4.1...v0.4.2) (2020-07-19)

### [0.4.1](https://github.com/pahud/cdk-spot-one/compare/v0.4.0...v0.4.1) (2020-07-19)


test/main.test.ts
import { App, Stack } from '@aws-cdk/core';
import '@aws-cdk/assert/jest';
import { SpotFleet } from '../src';

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

test('default cluster provision single ec2 instance', () => {
  const app = new App();
  const stack = new Stack(app, 'testing', { env: devEnv });
  new SpotFleet(stack, 'SpotFleet');
  expect(stack).toHaveResourceLike('AWS::EC2::SpotFleet', {
    SpotFleetRequestConfigData: {
      TargetCapacity: 1,
    },
  });
});

test('fleet with custom AMI ID comes with default linux userdata', () => {
  const app = new App();
  const stack = new Stack(app, 'testing', { env: devEnv });
  new SpotFleet(stack, 'SpotFleet', {
    customAmiId: 'ami-xxxxxx',
  });
  expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
    LaunchTemplateData: {
      UserData: {
        'Fn::Base64': '#!/bin/bash',
      },
    },
  });
});

test/index.test.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import { App, Stack, Duration } from '@aws-cdk/core';
import { SpotFleet, InstanceInterruptionBehavior } from '../src';
import '@aws-cdk/assert/jest';
import { BlockDuration } from '../src/spot';

describe('Spot Fleet tests', () => {
  let mockApp: App;
  let stack: Stack;

  beforeEach(() => {
    mockApp = new App();
    stack = new Stack(mockApp, 'testing-stack');
  });

  test('create the HTTP API', () => {
    const fleet = new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.SIX_HOURS,
      instanceInterruptionBehavior: InstanceInterruptionBehavior.HIBERNATE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      eipAllocationId: 'eipalloc-0d1bc6d85895a5410',
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
    });
    // fleet to expire after 6 hours
    fleet.expireAfter(Duration.hours(6));
    fleet.defaultSecurityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(80));
    expect(stack).toHaveResource('AWS::EC2::SpotFleet');
    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        InstanceMarketOptions: {
          MarketType: 'spot',
          SpotOptions: {
            BlockDurationMinutes: 360,
            InstanceInterruptionBehavior: 'hibernate',
          },
        },
      },
    });
    expect(stack).toHaveResource('AWS::EC2::SecurityGroup', {
      SecurityGroupIngress: [
        {
          CidrIp: '0.0.0.0/0',
          Description: 'from 0.0.0.0/0:22',
          FromPort: 22,
          IpProtocol: 'tcp',
          ToPort: 22,
        },
        {
          CidrIp: '0.0.0.0/0',
          Description: 'from 0.0.0.0/0:80',
          FromPort: 80,
          IpProtocol: 'tcp',
          ToPort: 80,
        },
      ],
    });
  });

  test('feet with custom security group that only allow http', () => {
    const securityGroup = new ec2.SecurityGroup(stack, 'Custom Security Group', {
      vpc: new ec2.Vpc(stack, 'VPC'),
    });
    securityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(80));

    new SpotFleet(stack, 'SpotFleet', {
      securityGroup,
    });

    expect(stack).toHaveResource('AWS::EC2::SecurityGroup', {
      SecurityGroupIngress: [
        {
          CidrIp: '0.0.0.0/0',
          Description: 'from 0.0.0.0/0:80',
          FromPort: 80,
          IpProtocol: 'tcp',
          ToPort: 80,
        },
      ],
    });
  });

  test('feet with custom instance role', () => {
    const anotherStack = new Stack(mockApp, 'another-stack');

    new SpotFleet(stack, 'SpotFleet', {
      instanceRole: new iam.Role(anotherStack, 'Custom Role', {
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      }),
    });

    expect(stack).not.toHaveResourceLike('AWS::IAM::Role', {
      ManagedPolicyArns: [
        'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
      ],
    });
    expect(anotherStack).toHaveResourceLike('AWS::IAM::Role', {
      ManagedPolicyArns: [
        'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
      ],
    });
  });

  test('feet without custom instance role comes with default role', () => {
    new SpotFleet(stack, 'SpotFleet', {});

    expect(stack).toHaveResourceLike('AWS::IAM::Role', {
      ManagedPolicyArns: [
        'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
      ],
    });
  });

  test('support additional user data', () => {
    new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.SIX_HOURS,
      instanceInterruptionBehavior: InstanceInterruptionBehavior.HIBERNATE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
      additionalUserData: [
        'mycommand1',
        'mycommand2 arg1',
      ],
    });

    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        UserData: {
          'Fn::Base64': '#!/bin/bash\nyum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm\nyum install -y docker\nusermod -aG docker ec2-user\nusermod -aG docker ssm-user\nservice docker start\nmycommand1\nmycommand2 arg1',
        },
      },
    });
  });

  test('long time spot fleet', () => {
    new SpotFleet(stack, 'SpotFleet', {
      targetCapacity: 1,
      blockDuration: BlockDuration.NONE,
      defaultInstanceType: new ec2.InstanceType('t3.large'),
      vpcSubnet: {
        subnetType: ec2.SubnetType.PUBLIC,
      },
      terminateInstancesWithExpiration: true,
    });

    expect(stack).toHaveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: {
        InstanceMarketOptions: {
          MarketType: 'spot',
          SpotOptions: {
            InstanceInterruptionBehavior: 'terminate',
          },
        },
      },
    });
  });
});


test/integ.default.ts
import { SpotFleet, InstanceInterruptionBehavior } from '../src/index';
import * as ec2 from '@aws-cdk/aws-ec2';
import { App, Stack, Duration } from '@aws-cdk/core';

const app = new App();

const env = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new Stack(app, 'SpotFleetStack', { env });

const fleet = new SpotFleet(stack, 'SpotFleet', {
  targetCapacity: 1,
  instanceInterruptionBehavior: InstanceInterruptionBehavior.HIBERNATE,
  defaultInstanceType: new ec2.InstanceType('t3.large'),
  eipAllocationId: 'eipalloc-0d1bc6d85895a5410',
  vpcSubnet: {
    subnetType: ec2.SubnetType.PUBLIC,
  },
  terminateInstancesWithExpiration: true,
})

// allocate this fleet for 6 hours
fleet.expireAfter(Duration.hours(6))









src/spot.ts
import * as path from 'path';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import { Construct, Resource, ResourceProps, PhysicalName, Stack, Fn, CfnOutput, Duration, Lazy, CustomResource, Token } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';

const DEFAULT_INSTANCE_TYPE = 't3.large';

export class VpcProvider extends Stack {
  public static getOrCreate(scope: Construct) {
    const stack = Stack.of(scope);
    return stack.node.tryGetContext('use_default_vpc') === '1' ?
      ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true }) :
      stack.node.tryGetContext('use_vpc_id') ?
        ec2.Vpc.fromLookup(stack, 'Vpc', { vpcId: stack.node.tryGetContext('use_vpc_id') }) :
        new ec2.Vpc(stack, 'Vpc', { maxAzs: 3, natGateways: 1 });
  }
}

export enum BlockDuration {
  ONE_HOUR = 60,
  TWO_HOURS = 120,
  THREE_HOURS = 180,
  FOUR_HOURS = 240,
  FIVE_HOURS = 300,
  SIX_HOURS = 360,
  NONE = 0,
}

/**
 * Whether the worker nodes should support GPU or just standard instances
 */
export enum NodeType {
  /**
   * Standard instances
   */
  STANDARD = 'Standard',

  /**
   * GPU instances
   */
  GPU = 'GPU',

  /**
   * Inferentia instances
   */
  INFERENTIA = 'INFERENTIA',

  /**
   * ARM instances
   */
  ARM = 'ARM',


}

export enum InstanceInterruptionBehavior {
  HIBERNATE = 'hibernate',
  STOP = 'stop',
  TERMINATE = 'terminate'
}

export interface SpotFleetLaunchTemplateConfig {
  readonly spotfleet: SpotFleet;
  readonly launchTemplate: ILaunchtemplate;
}

export interface ILaunchtemplate {
  bind(spotfleet: SpotFleet): SpotFleetLaunchTemplateConfig;
}

export class LaunchTemplate implements ILaunchtemplate {
  public bind(spotfleet: SpotFleet): SpotFleetLaunchTemplateConfig {
    return {
      spotfleet,
      launchTemplate: this,
    };
  }
}

export interface BaseSpotFleetProps extends ResourceProps {
  /**
   * VPC for the spot fleet
   *
   * @default - new VPC will be created
   *
   */
  readonly vpc?: ec2.IVpc;

  /**
   * default EC2 instance type
   *
   * @default - t3.large
   */
  readonly defaultInstanceType?: ec2.InstanceType;

  /**
   * reservce the spot instance as spot block with defined duration
   *
   * @default - BlockDuration.ONE_HOUR
   */
  readonly blockDuration?: BlockDuration;

  /**
   * The behavior when a Spot Instance is interrupted
   *
   * @default - InstanceInterruptionBehavior.TERMINATE
   */
  readonly instanceInterruptionBehavior?: InstanceInterruptionBehavior;

  /**
   * IAM role for the spot instance
   */
  readonly instanceRole?: iam.Role;

  /**
   * number of the target capacity
   *
   * @default - 1
   */
  readonly targetCapacity?: number;

  /**
   * the time when the spot fleet allocation starts
   *
   * @default - no expiration
   */
  readonly validFrom?: string;

  /**
   * the time when the spot fleet allocation expires
   *
   * @default - no expiration
   */
  readonly validUntil?: string;

  /**
   * terminate the instance when the allocation is expired
   *
   * @default - true
   */
  readonly terminateInstancesWithExpiration?: boolean;

  /**
   * custom AMI ID
   *
   * @default - The latest Amaozn Linux 2 AMI ID
   */
  readonly customAmiId?: string;

  /**
   * VPC subnet for the spot fleet
   *
   * @default - public subnet
   */
  readonly vpcSubnet?: ec2.SubnetSelection;

  /**
   * Security group for the spot fleet
   *
   * @default - allows TCP 22 SSH ingress rule
   */
  readonly securityGroup?: ec2.SecurityGroup;

  /**
   * SSH key name
   *
   * @default - no ssh key will be assigned
   */
  readonly keyName?: string;

  /**
   * Additional commands for user data
   *
   * @default - no additional user data
   */
  readonly additionalUserData?: string[];

  /**
   * blockDeviceMappings for config instance.
   *
   * @default - from ami config.
   */
  readonly blockDeviceMappings?: ec2.CfnLaunchTemplate.BlockDeviceMappingProperty[] | undefined;
}

export interface SpotFleetProps extends BaseSpotFleetProps {
  /**
   * Launch template for the spot fleet
   */
  readonly launchTemplate?: ILaunchtemplate;

  /**
   * Allocation ID for your existing Elastic IP Address.
   *
   * @defalt new EIP and its association will be created for the first instance in this spot fleet
   */
  readonly eipAllocationId?: string;

}

export class SpotFleet extends Resource {
  readonly instanceRole: iam.IRole;
  readonly defaultInstanceType: ec2.InstanceType;
  readonly targetCapacity?: number;
  readonly spotFleetId: string;
  readonly launchTemplate: ILaunchtemplate;
  readonly vpc: ec2.IVpc;
  /**
   * the first instance id in this fleet
   */
  readonly instanceId: string;
  /**
   * instance type of the first instance in this fleet
   */
  readonly instanceType: string;
  /**
   * SpotFleetRequestId for this spot fleet
   */
  readonly spotFleetRequestId: string;
  /**
   * The time when the the fleet allocation will expire
   */
  private validUntil?: string;
  /**
   * The default security group of the instance, which only allows TCP 22 SSH ingress rule.
   */
  public readonly defaultSecurityGroup: ec2.ISecurityGroup;

  constructor(scope: Construct, id: string, props: SpotFleetProps = {}) {
    super(scope, id, props);

    this.spotFleetId = id;
    this.launchTemplate = props.launchTemplate ?? new LaunchTemplate();
    this.targetCapacity = props.targetCapacity ?? 1;
    this.defaultInstanceType = props.defaultInstanceType ?? new ec2.InstanceType(DEFAULT_INSTANCE_TYPE);
    this.validUntil = props.validUntil;
    this.vpc = props.vpc ?? new ec2.Vpc(this, 'VPC', { maxAzs: 3, natGateways: 1 });
    this.defaultSecurityGroup = props.securityGroup || this.createSecurityGroup();
    this.instanceRole = props.instanceRole || this.createInstanceRole();
    this.instanceRole.addManagedPolicy({
      managedPolicyArn: 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
    });
    const stack = Stack.of(this);

    const instanceProfile = new iam.CfnInstanceProfile(this, 'InstanceProfile', {
      roles: [this.instanceRole.roleName],
    });

    this.defaultInstanceType = props.defaultInstanceType ?? new ec2.InstanceType(DEFAULT_INSTANCE_TYPE);

    const imageId = props.customAmiId ??
      ec2.MachineImage.latestAmazonLinux({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
        cpuType: nodeTypeForInstanceType(this.defaultInstanceType) === NodeType.ARM ? ec2.AmazonLinuxCpuType.ARM_64 : undefined,
      }).getImage(this).imageId;

    const userData = ec2.UserData.forLinux();

    /**
     * If not using custom AMI, we use amazon linux 2 and install the SSM agent and enable docker by default.
     * Otherwise, we simply do nothing here.
     */
    if (props.customAmiId == undefined) {
      userData.addCommands(
        'yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm',
        'yum install -y docker',
        'usermod -aG docker ec2-user',
        'usermod -aG docker ssm-user',
        'service docker start',
      );
    }

    if (props.additionalUserData) {userData.addCommands(...props.additionalUserData);}
    const lt = new ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        imageId,
        instanceType: this.defaultInstanceType.toString(),
        userData: Fn.base64(userData.render()),
        keyName: props.keyName,
        tagSpecifications: [
          {
            resourceType: 'instance',
            tags: [
              {
                key: 'Name',
                value: `${Stack.of(this).stackName}/spotFleet/${this.spotFleetId}`,
              },
            ],
          },
        ],
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        instanceMarketOptions: {
          marketType: 'spot',
          spotOptions: {
            blockDurationMinutes: (props.blockDuration == BlockDuration.NONE) ? undefined : (props.blockDuration ?? BlockDuration.ONE_HOUR),
            instanceInterruptionBehavior: props.instanceInterruptionBehavior ?? InstanceInterruptionBehavior.TERMINATE,
          },
        },
        securityGroupIds: this.defaultSecurityGroup.connections.securityGroups.map(m => m.securityGroupId),
        iamInstanceProfile: {
          arn: instanceProfile.attrArn,
        },
      },
    });

    const spotFleetRole = new iam.Role(this, 'FleetRole', {
      assumedBy: new iam.ServicePrincipal('spotfleet.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonEC2SpotFleetTaggingRole'),
      ],
    });

    const vpcSubnetSelection = props.vpcSubnet ?? {
      subnetType: ec2.SubnetType.PUBLIC,
    };
    const subnetConfig = this.vpc.selectSubnets(vpcSubnetSelection).subnets.map(s => ({
      subnetId: s.subnetId,
    }));
    const cfnSpotFleet = new ec2.CfnSpotFleet(this, id, {
      spotFleetRequestConfigData: {
        launchTemplateConfigs: [
          {
            launchTemplateSpecification: {
              launchTemplateId: lt.ref,
              version: lt.attrLatestVersionNumber,
            },
            overrides: subnetConfig,
          },
        ],
        iamFleetRole: spotFleetRole.roleArn,
        targetCapacity: props.targetCapacity ?? 1,
        validFrom: props.validFrom,
        validUntil: Lazy.stringValue({ produce: () => this.validUntil }),
        terminateInstancesWithExpiration: props.terminateInstancesWithExpiration ?? true,
      },
    });
    new CfnOutput(stack, 'SpotFleetId', { value: cfnSpotFleet.ref });
    const onEvent = new lambda.Function(this, 'OnEvent', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../eip-handler')),
      handler: 'index.on_event',
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(60),
    });

    const isComplete = new lambda.Function(this, 'IsComplete', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../eip-handler')),
      handler: 'index.is_complete',
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(60),
      role: onEvent.role,
    });

    const myProvider = new cr.Provider(this, 'MyProvider', {
      onEventHandler: onEvent,
      isCompleteHandler: isComplete, // optional async "waiter"
      logRetention: logs.RetentionDays.ONE_DAY, // default is INFINITE
    });

    onEvent.addToRolePolicy(new iam.PolicyStatement({
      actions: ['ec2:DescribeSpotFleetInstances'],
      resources: ['*'],
    }));

    const fleetInstances = new CustomResource(this, 'SpotFleetInstances', {
      serviceToken: myProvider.serviceToken,
      properties: {
        SpotFleetRequestId: cfnSpotFleet.ref,
      },
    });

    fleetInstances.node.addDependency(cfnSpotFleet);

    this.instanceId = Token.asString(fleetInstances.getAtt('InstanceId'));
    this.instanceType = Token.asString(fleetInstances.getAtt('InstanceType'));
    this.spotFleetRequestId = Token.asString(fleetInstances.getAtt('SpotInstanceRequestId'));

    new CfnOutput(stack, 'InstanceId', { value: this.instanceId });

    // EIP association
    if (props.eipAllocationId) {
      new ec2.CfnEIPAssociation(this, 'EipAssocation', {
        allocationId: props.eipAllocationId,
        instanceId: this.instanceId,
      });
    } else {
      new ec2.CfnEIP(this, 'EIP', {
        instanceId: this.instanceId,
      });
    }
  }
  public expireAfter(duration: Duration) {
    const date = new Date();
    date.setSeconds(date.getSeconds() + duration.toSeconds());
    this.validUntil = date.toISOString();
  }

  private createInstanceRole(): iam.Role {
    return new iam.Role(this, 'InstanceRole', {
      roleName: PhysicalName.GENERATE_IF_NEEDED,
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
    });
  }

  private createSecurityGroup(): ec2.SecurityGroup {
    const securityGroup = new ec2.SecurityGroup(this, 'SpotFleetSg', {
      vpc: this.vpc,
    });
    securityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(22));
    return securityGroup;
  }
}


const GRAVITON_INSTANCETYPES = ['a1'];
const GRAVITON2_INSTANCETYPES = ['c6g', 'm6g', 'r6g'];
const GPU_INSTANCETYPES = ['p2', 'p3', 'g4'];
const INFERENTIA_INSTANCETYPES = ['inf1'];

function nodeTypeForInstanceType(instanceType: ec2.InstanceType) {
  return GPU_INSTANCETYPES.includes(instanceType.toString().substring(0, 2)) ? NodeType.GPU :
    INFERENTIA_INSTANCETYPES.includes(instanceType.toString().substring(0, 4)) ? NodeType.INFERENTIA :
      GRAVITON2_INSTANCETYPES.includes(instanceType.toString().substring(0, 3)) ? NodeType.ARM :
        GRAVITON_INSTANCETYPES.includes(instanceType.toString().substring(0, 2)) ? NodeType.ARM :
          NodeType.STANDARD;
}
